sapicore/utils/io.py

```diff
@@ -0,0 +1,270 @@
+
+    Parameters
+    ----------
+    level: int
+        Logger level, `INFO` by default.
+
+    stream:
+        Destination stream, `stdout` by default.
+
+    file:
+        Destination file, `None` by default.
+
+    formatting: str
+        Message formatting.
+
+    """
+    logging.basicConfig(
+        level=level,
+        format=formatting,
+        handlers=[logging.StreamHandler(stream)] + ([logging.FileHandler(file)] if file else []),
+    )
+
+
+def load_apply_config(configuration: str, apply_to: object = None) -> dict:
+    # parse configuration from YAML if given and use resulting dictionary to initialize attributes.
+    content = load_config(None, configuration) if os.path.exists(configuration) else None
+
+    if not content:
+        raise FileNotFoundError(f"Could not find a configuration YAML at {os.path.realpath(configuration)}")
+
+    else:
+        # apply the configuration to this pipeline object.
+        configuration = content
+
+        if apply_to:
+            apply_config(apply_to, configuration)
+
+    return configuration
+
+
+def ensure_dir(path: str = None) -> str:
+    """Ensures that a path is created if parts thereof do not exist yet and returns it.
+
+    Parameters
+    ----------
+    path: str
+        File path to be verified or created.
+
+    Returns
+    -------
+    path: str
+        The same file path, so it can be used in the calling context (e.g., if :func:`ensure_dir`
+        wraps an assignment).
+
+    """
+    os.makedirs(path, exist_ok=True)
+    return path
+
+
+def flatten(lst: list[list]) -> list:
+    """Flattens a list."""
+    return [item for sublist in lst for item in sublist]
+
+
+def load_yaml(path: str) -> dict:
+    """Parses the YAML at `path`, returning a dictionary."""
+    return yaml.safe_load(Path(path).read_text())
+
+
+def save_yaml(data: dict, path: str):
+    """Saves a dictionary `data` to the YAML file specified by `path`."""
+    with open(path, "w") as f:
+        yaml.dump(data, f, default_flow_style=False)
+
+
+def plot_tensorboard(configuration: dict, data_dir: str, tb_dir: str):
+    """Loads simulation output data from disk and writes it to tensorboard for visual inspection."""
+    writer = TensorboardWriter(log_dir=tb_dir)
+    with alive_bar(total=len(os.listdir(data_dir)), force_tty=True) as bar:
+        logging.info("Writing tensorboard data for visualization.")
+        for file in os.listdir(data_dir):
+            if not file.endswith(".h5"):
+                continue
+
+            # write what was asked.
+            tensorboard_settings = configuration.get("simulation", {}).get("tensorboard", 0)
+            if isinstance(tensorboard_settings, list):
+                for task in tensorboard_settings:
+                    for attr, settings in task.items():
+                        # read only requested key from file.
+                        data = HDFData(path=os.path.join(data_dir, file), key=attr)
```

Comment:
There's something erking me here. Because you have a folder with logged files and a single config dict. The dict applies to all files in the folder? So for all the files we extract the same properties? Would we want to maybe have a per file config dict?

You're using `plot_tensorboard` only in generic pipeline rn so I understand that all files would be processed the same way. But, as a stand alone function, maybe it should instead accept a h5 filename, a list of attrs and settings and have the pipeline do the folder crawling etc. Partially it's also because `HDFData` accepts a buffer, so why not re-use it if we're reading multiple attrs from the same file. Or, why have it accept a buffer? If the same attr can be specified multiple times, then `TensorboardWriter` should be able to deal with that making it unique. If it can't, like it currently seems to be the case, the not just do it all in one `HDFData`?

But anyway that can all be left for later.

Reply:
> you have a folder with logged files and a single config dict

Note that the master config file contains references to neuron and synapse YAMLs (each corresponding to one .h5 log file). We should copy those to the log directory as well for completeness--thanks for bringing this up.

Good points regarding `HDFData`. Unraveling and refactoring that will have to wait for a future patch; there are some nontrivial design decisions involved.